internal/racetime/api.go

```diff
@@ -69,19 +79,114 @@ type CategoryResponse struct {
 	CurrentRaces      []RaceData `json:"current_races"`
 }
 
+type LeaderboardsResponse struct {
+	Goal      string `json:"goal"`
+	NumRanked string `json:"num_ranked"`
+	Rankings  struct {
+		User         string `json:"user"`
+		Place        string `json:"place"`
+		PlaceOrdinal string `json:"place_ordinal"`
+		Score        int    `json:"score"`
+		TimesRaced   string `json:"times_raced"`
+	}
+}
+
+type UserSearchParameters struct {
+	Name string
+	Scrim string
+	Term string
+}
+
 // CategoryDetail fetches the race data of a specific racetime.gg category
 func CategoryDetail(c config.Racetime, category string) (*CategoryResponse, error) {
-	resp, err := http.Get(fmt.Sprintf("%s/%s/data", c.URL, category))
+	content, err := Get(c, fmt.Sprintf("%s/data", category), nil)
+
+	var cr *CategoryResponse
+	json.Unmarshal([]byte(content), cr)
 	if err != nil {
 		return nil, err
 	}
-	defer resp.Body.Close()
 
-	var cr *CategoryResponse
-	err = json.NewDecoder(resp.Body).Decode(cr)
+	return cr, nil
+}
+
+func CategoryLeaderboards(c config.Racetime, category string) (*LeaderboardsResponse, error) {
+	content, err := Get(c, fmt.Sprintf("%s/leaderboards/data", category), nil)
+
+	var cl *LeaderboardsResponse
+	json.Unmarshal([]byte(content), cl)
 	if err != nil {
 		return nil, err
 	}
 
-	return cr, nil
+	return cl, nil
 }
+
+// RaceDetail fetches the race data of a specific racetime.gg race
+func RaceDetail(c config.Racetime, category string, race string) (*RaceData, error) {
+	content, err := Get(c, fmt.Sprintf("%s/%s/data", category, race), nil)
+
+	var rd *RaceData
+	json.Unmarshal([]byte(content), rd)
+	if err != nil {
+		return nil, err
+	}
+
+	return rd, nil
+}
+
+func PastUserRaces(c config.Racetime, user string, showEntrants bool, page int){
+	content, err := Get(c, fmt.Sprintf("user/%s/races/data", user), nil)
+
+	var pur *RaceData
+	json.Unmarshal([]byte(content), pur)
+	if err != nil {
+	}
+}
+
+func UserSearch(c config.Racetime, user string){
+	parameters := UserSearchParameters{
+		Name: "colfra",
+	}
+
+	m := structs.Map(parameters)
+	content, _ := Get(c, "user/search", m)
+	var userSearchResults UserDataResponse
+	json.Unmarshal([]byte(content), &userSearchResults)
+	log.Println(userSearchResults.Results[0].ID)
+}
+
+func GetTest(){
+	var test = `{"results": [{"id": "5BRGVMd30E368Lzv", "full_name": "colfra", "name": "colfra", "discriminator": null, "url": "/user/5BRGVMd30E368Lzv", "avatar": null, "pronouns": null, "flair": "staff", "twitch_name": null, "twitch_display_name": null, "twitch_channel": null, "can_moderate": false}]}`;
+	jd := json.NewDecoder(strings.NewReader(test))
+	var userResults UserDataResponse
+	jd.Decode(&userResults)
+	log.Println(userResults.Results[0].ID)
+}
+
+func Get(c config.Racetime, endpoint string, parameters map[string]interface{}) (string, error) {
```

Comment:
Instead of having `parameters` be a `map[string]interface{}`, let's instead just accept a `url.Values` type for parameters. That way we can just assign that via `requestUrl.RawQuery = parameters.Encode()`.

This would avoid having to inherit any of this iteration complexity.

Reply:
I'll also say, let's just return `([]byte, error)` instead of casting it to a string. `[]byte` is the most common type for this sort of thing and that way we don't have to re-cast/un-cast later for json decoding

Reply:
An alternative to this I actually like is to return an `io.ReadCloser` instead, then just `return resp.Body, nil` at the end. This returns an interface where we expect the caller to read/close from the body, and we don't have to load those bytes fully into memory before this function can return. This also works nicely because you can decode json directly from an `io.Reader` (of which `io.ReadCloser` embeds)